magma/backend/coreir_.py

```diff
@@ -69,21 +70,26 @@ def magma_port_to_coreir(port):
 
     return select.replace("[", ".").replace("]", "")
 
+# Singleton context meant to be used with coreir/magma code
+@singleton
+class CoreIRContextSingleton:
+    __instance = None
 
-magma_coreir_context = coreir.Context()  # Singleton context meant to be used with coreir/magma code
-def __reset_context():
-    """
-    Testing hook so every test has a fresh context
-    """
-    global magma_coreir_context
-    magma_coreir_context = coreir.Context()
+    def get_instance(self):
+        return self.__instance
 
+    def reset_instance(self):
+        self.__instance = coreir.Context()
+
+    def __init__(self):
+        self.__instance = coreir.Context()
+CoreIRContextSingleton()
 
 class CoreIRBackend:
     context_to_modules_map = {}
     def __init__(self, context=None):
         if context is None:
-            context = magma_coreir_context
+            context = CoreIRContextSingleton().get_instance()
```

Comment:
Since we're phasing out the notion of different contexts, perhaps we can add an assert here that the passed in context is the same as the singleton context? Alternatively, we could do a check and issue a warning (not sure how much code out there is relying on the ability to use their own coreir context)

Reply:
I think it's best to avoid an argument that must have one value at all times. I don't want to crash if the context is not the singleton context. There also may be uses for a non-singleton CoreIR context. 

I've added in a warning if using the non-singleton context and a flag for disabling the warning.